test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,258 @@
+            "(Discovered while processing\n{}.)").format(functional_full_name, pprint.pformat(test_params_dict))
+
+        for device in devices:
+            test_params = process_test_params_for_functional(
+                test_params_dict=test_params_dict,
+                device=device,
+                test_instance_class=test_instance_class,
+            )
+            test_name = 'test_torch_nn_functional_{}'.format(test_params.functional_variant_name)
+            unit_test_class.functional_test_params_map[test_name] = test_params
+
+            def test_fn(self):
+                test_torch_nn_functional_variant(
+                    unit_test_class=self, test_params=unit_test_class.functional_test_params_map[self._testMethodName])
+
+            test_fn = decorate_test_fn(
+                test_fn=test_fn,
+                test_cpp_api_parity=test_params_dict.get('test_cpp_api_parity', True),
+                test_cuda=test_params_dict.get('test_cuda', True),
+                has_impl_parity=parity_table['torch::nn::functional'][functional_full_name][0] and
+                test_params_dict.get('has_parity', True),
+                device=device)
+
+            add_test(unit_test_class, test_name, test_fn)
+
+def write_tests_to_test_class(unit_test_class, test_params_dicts, test_instance_class, parity_table, devices):
+    add_torch_nn_functional_impl_parity_tests(
+        parity_table=parity_table,
+        unit_test_class=unit_test_class,
+        test_params_dicts=test_params_dicts,
+        test_instance_class=test_instance_class,
+        devices=devices)
+
+def generate_test_cpp_sources(test_params, template):
+    cpp_args_construction_stmts, _ = compute_cpp_args_construction_stmts_and_forward_arg_symbols(test_params)
+
+    test_cpp_sources = template.substitute(
+        functional_variant_name=test_params.functional_variant_name,
+        cpp_args_construction_stmts=";\n  ".join(cpp_args_construction_stmts),
+        cpp_function_call=test_params.cpp_function_call,
+    )
+    return test_cpp_sources
+
+# Build all C++ tests together, instead of once per test.
+def build_cpp_tests(unit_test_class, print_cpp_source=False):
+    assert len(unit_test_class.functional_test_params_map) > 0
+    cpp_sources = TORCH_NN_COMMON_TEST_HARNESS
+    functions = []
+    functionals_added_cpp_sources = set()
+    for test_name, test_params in unit_test_class.functional_test_params_map.items():
+        if test_params.functional_name not in functionals_added_cpp_sources:
+            cpp_sources += cpp_api_parity.functional_cpp_sources.get(test_params.functional_name, '')
+            functionals_added_cpp_sources.add(test_params.functional_name)
+        cpp_sources += generate_test_cpp_sources(test_params=test_params, template=TORCH_NN_FUNCTIONAL_TEST_FORWARD)
+        functions.append('{}_{}'.format(test_params.functional_variant_name, 'test_forward'))
```

Comment:
Could we just add a `test_name` to the `test_params` named tuple?

Reply:
Sounds great and I think we should do it, and I think we can do it at a later time when we add other test functions (such as `test_constructor_args` or `test_init`) for each module test dict.